astropy/io/ascii/cds.py

```diff
@@ -164,14 +210,437 @@ def get_cols(self, lines):
+            # for values in Scientific notation, when writing that Format.
+            col.fortran_format = fformat + str(col.meta.size) + "." + str(maxprec)
+            col.format = str(col.meta.size) + "." + str(maxdec) + "e"
+        else:
+            col.meta.size = maxent + maxdec + 1
+            if sign:
+                col.meta.size += 1
+            col.fortran_format = fformat + str(col.meta.size) + "." + str(maxdec)
+            col.format = col.fortran_format[1:] + "f"
 
-class CdsData(core.BaseData):
+    def write_byte_by_byte(self):
+        """
+        Writes the Byte-By-Byte description of the table.
+
+        Columns that are `astropy.coordinates.SkyCoord` or `astropy.time.TimeSeries`
+        objects or columns with values that are such objects are recognized as such,
+        and some predefined labels and description is used for them.
+        See the Vizier CDS Standard documentation in the link below for more details
+        on these. An example Byte-By-Byte table is shown here.
+
+        See: http://vizier.u-strasbg.fr/doc/catstd-3.1.htx
+
+        Example::
+
+        --------------------------------------------------------------------------------
+        Byte-by-byte Description of file: table.dat
+        --------------------------------------------------------------------------------
+        Bytes Format Units  Label     Explanations
+        --------------------------------------------------------------------------------
+         1- 8   A8     ---    names    Description of names
+        10-14   E5.1   ---    e       [-3160000.0/0.01] Description of e
+        16-23   F8.5   ---    d       [22.25/27.25] Description of d
+        25-31   E7.1   ---    s       [-9e+34/2.0] Description of s
+        33-35   I3     ---    i       [-30/67] Description of i
+        37-39   F3.1   ---    sameF   [5.0/5.0] Description of sameF
+        41-42   I2     ---    sameI   [20] Description of sameI
+        44-47   F4.1   h      RAh      Right Ascension (hour)
+        49-51   F3.1   min    RAm      Right Ascension (minute)
+        53-70   F18.15 s      RAs      Right Ascension (second)
+           72   A1     ---    DE-     Sign of Declination
+        73-76   F5.1   deg    DEd      Declination (degree)
+        78-81   F4.1   arcmin DEm      Declination (arcmin)
+        83-98   F16.13 arcsec DEs      Declination (arcsec)
+
+        --------------------------------------------------------------------------------
+        """
+        # For columns that are instances of ``SkyCoord`` and ``TimeSeries`` classes,
+        # or whose values are objects of these classes.
+        for i, col in enumerate(self.cols):
+            # If col is a ``Column`` object but its values are ``SkyCoord`` objects,
+            # convert the whole column to ``SkyCoord`` object, which helps in method
+            # applications.
+            if not isinstance(col, SkyCoord) and isinstance(col[0], SkyCoord):
```

Comment:
We talked a little about this during the last meeting. I thought that it would be good to make the writer flexible enough to write a ``SkyCoord`` column if all elements of the column are ``SkyCoord`` -- however, what you've pointed out is an important case to deal with. Would it be better to do a ``try-except`` with ``col = SkyCoord(col)`` and raise a warning  if it fails along with skipping the column? Or raising an error to prompt the user to remove the column or fix it themselves? I am more inclined towards the first option since it writes something while warning, but happy to hear more thoughts!

Reply:
There is also a question what happens if a table has two SkyCoord columns, say, I cross-matched two catalogs based on their coordinates and both of the original coordinate columns are still in the table. The current code would convert the first column and then delete the original, and then later convert the second column, overwrite the names "RAh", etc. so that all information on the first column is silently lost. I think we need either an error, a warning, or some way for the user to select which column is the one that should be converted to "RAH", ...

Reply:
> I think we need either an error, a warning, or some way for the user to select which column is the one that should be converted to "RAH", ...

We also discussed about this case previously and opted to have a note in the documentation that currently tables with only one single Coordinate column are supported. There's no separate criteria in the code to deal with situations of two or more coordinate columns. In those cases, sometimes what you say will happen or when one column is in Ecliptic/Galactic and other is in RA/DEC, some random warnings are generated.
Let's talk more about it in the meeting.

Reply:
I agree we don't have to support them (yet), but we do need to make sure that nothing bad happens silently. I suggest something like:
```
if ['Rah', 'RAm', ... 'ELAT'] in self.colnames:
    warnings.warn(f'Table already has coordinate system in CDS/MRT-syle columns. Skipping column {col}')
```
Two lines of code, easy to change later, but we make sure that limitations are written out, which is important before we can merge this.
(Code above might not work exactly as written. Not sure is self.colnames is defined at this point.)

Reply:
> We talked a little about this during the last meeting. I thought that it would be good to make the writer flexible enough to write a `SkyCoord` column if all elements of the column are `SkyCoord` -- however, what you've pointed out is an important case to deal with. Would it be better to do a `try-except` with `col = SkyCoord(col)`, or explicitly check if all elements of the column are ``SkyCoord`, and then raise a warning + skip the column if it fails? Else, an option is to raise an error and prompt the user to remove the column/fix it themselves? I am more inclined towards the first option since it writes something while warning, but happy to hear more thoughts!

I have implemented a ``try-except`` for the esoteric cases where a column has first value as a ``SkyCoord`` object but not others. However, these types of columns cannot be parsed properly for generation of the Byte-By-Byte, even when they are `Column` objects, because consistency across the whole column is expected. Also, that singleton ``SkyCoord`` object or others if present in the column, will need to be converted to float/string values. The data would in any case be wrong.
These columns are converted to ``Column`` object and then skipped from further editing. However, they may still result in failure.

Reply:
I checked with other `mix-in` type columns, for instance, the various coordinate Representations. The `write_byte_by_byte` function fails to interpret the `size` and `format` for all such columns, because their values are of `object` type and not any of the `float`, `int` or `string` types for which functions are provided.

For these `mix-in` columns and columns with partial `SkyCoord` values, I first make a `Column` object and convert each of the column values to `str`. Henceforth, those columns are then treated as normal string columns, with the `label` been put as `Unknown`.

The earlier traceback from the writer while writing these `mix-in` columns was as given below. Now, the table is written as given the tests `test_write_partial_SkyCoord_col` and `test_write_other_mixin_cols`.

```
(.virtualenvs) $ python3.8 cdsWriterAstropyRun.py check_otherMixInCols
Traceback (most recent call last):
  File "cdsWriterAstropyRun.py", line 540, in <module>
    func()
  File "cdsWriterAstropyRun.py", line 487, in check_otherMixInCols
    ascii.write(t, format='cds')
  File "/media/suyog/DATA/Python/astropy_fork/astropy/io/ascii/ui.py", line 845, in write
    lines = writer.write(table)
  File "/media/suyog/DATA/Python/astropy_fork/astropy/io/ascii/cds.py", line 845, in write
    return super().write(table)
  File "/media/suyog/DATA/Python/astropy_fork/astropy/io/ascii/core.py", line 1488, in write
    self.write_header(lines, table.meta)
  File "/media/suyog/DATA/Python/astropy_fork/astropy/io/ascii/core.py", line 1443, in write_header
    self.header.write(lines)
  File "/media/suyog/DATA/Python/astropy_fork/astropy/io/ascii/cds.py", line 633, in write
    'bytebybyte': self.write_byte_by_byte()})
  File "/media/suyog/DATA/Python/astropy_fork/astropy/io/ascii/cds.py", line 514, in write_byte_by_byte
    endb = col.meta.size + startb - 1
AttributeError: 'collections.OrderedDict' object has no attribute 'size'
```

Reply:
Maybe it's better to just skip any column of type "object"? We take care of mix-in coordiantes, we will take care of mix-in times and for all others users will need to do the conversion "by hand" before calling this writer. There is precendence - not every writer can write every column format, e.g. a csv table with multi-D columns won't work.

Reply:
> For these `mix-in` columns and columns with partial `SkyCoord` values, I first make a `Column` object and convert each of the column values to `str`. Henceforth, those columns are then treated as normal string columns, with the `label` been put as `Unknown`.

Uhm... But I have already add this for the `Time` and other `mix-in` columns. I personally thing it is better to output a table with string columns, than an `AttributeError` being raised later. When we have support for `Time` columns, I shall modify parts of this code. Let me push these new commits and then we can discuss further.

Reply:
It the same case for extra coordinate columns. They are skipped, but later another `AttributeError` comes up because the skipped column is still a `SkyCoord` object.

I would prefer to **issue warning** and **convert such col values to string**, so the writing can proceed ahead and **label all these cols as  _Unknown_**. The final output table is not useful, but the user will need to make modification anyway. I think it good to show up some output table in all situations.

Any corrections for all these standalone corner cases can be incorporated individually later on.

Reply:
A warning is fine, too. I just don't want to silently convert one column and overwrite the previous one without the user knowing that's happening.

Reply:
I agree that converting columns behind the scenes without the user knowing is __bad__. I like the idea of issuing a warning to let the user know that they need to convert the `object` columns by hand (except co-ordinate and time columns which you will deal with either in this PR or later), but still outputting something so that the user knows where the output is failing. However, the `Table` should ideally not be converted, i.e. a copy works well.